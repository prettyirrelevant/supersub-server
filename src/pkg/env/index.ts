import { z } from 'zod';

const zEnv = z.object({
  ENVIRONMENT: z.enum(['development', 'production']).default('development'),
  LOG_LEVEL: z.enum(['debug', 'info', 'warn', 'error']).default('debug'),
  ALCHEMY_WEBHOOK_SIGNING_KEY: z.string().min(1),
  DATABASE_URL: z.string().url().min(1),
  PORT: z.coerce.number().default(3456),
  PRIVY_APP_SECRET: z.string().min(1),
  ALCHEMY_API_KEY: z.string().min(1),
  REDIS_URL: z.string().url().min(1),
  RESEND_API_KEY: z.string().min(1),
  PRIVY_APP_ID: z.string().min(1),
  PRIVATE_KEY: z.string().min(1),
  SECRET_KEY: z.string().min(1),
});

const parsedEnv = zEnv.safeParse(process.env);
if (!parsedEnv.success) {
  throw new Error(`Unable to parse environment variables ${parsedEnv.error}`);
}

export const config = parsedEnv.data;
