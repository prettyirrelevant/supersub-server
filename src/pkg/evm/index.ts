export { SUBSCRIPTION_PLUGIN_INIT_BLOCK, SUBSCRIPTION_PLUGIN_ADDRESS, CHUNK_SIZE } from './constants';
export { getMultiOwnerModularAccountAddresses, getEvmHttpClient } from './utils';
export { SubscriptionPluginAbi } from './abis';
