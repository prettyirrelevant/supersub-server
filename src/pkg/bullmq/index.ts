import { Worker, Queue } from 'bullmq';
import path from 'path';

import { logger } from '~/pkg/logging';
import { redis } from '~/pkg/redis';
import { config } from '~/pkg/env';

const queue = new Queue('supersub-queue', { connection: redis });

const addJobsToQueue = async () => {
  // Runs every minute
  await queue.add('enrich-tokens', {}, { repeat: { pattern: '* * * * *' } });

  // Runs every 12 hours (0th minute of every 12th hour)
  await queue.add('renew-subscriptions', {}, { repeat: { pattern: '0 */12 * * *' } });

  // Runs every five(5) minutes
  await queue.add('fetch-smart-accounts', {}, { repeat: { every: 300_000 } });

  // Runs every 30 seconds
  await queue.add('index-subscription-plugin-events', {}, { repeat: { every: 30_000 } });

  // Runs every 24 hours (daily at midnight)
  await queue.add('upcoming-subscriptions-renewal-reminders', {}, { repeat: { pattern: '0 0 * * *' } });
};

addJobsToQueue();

const sandboxedWorkerPath = () => {
  return config.ENVIRONMENT === 'development'
    ? path.join(path.dirname(path.dirname(path.dirname(__filename))), 'workers', `sandboxed.ts`)
    : path.join(path.dirname(__filename), 'workers', 'sandboxed.js');
};

const worker = new Worker('supersub-queue', sandboxedWorkerPath(), {
  connection: redis,
});

process.on('SIGTERM', async () => {
  logger.info('SIGTERM signal received: closing queues');

  await worker.close();
  logger.info('All closed');
});

export { queue };
