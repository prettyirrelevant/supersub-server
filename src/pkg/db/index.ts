import { PrismaClient } from '@prisma/client';

import { config } from '~/pkg/env';

// add prisma to the NodeJS global type
interface CustomNodeJsGlobal extends Global {
  prisma: PrismaClient;
}

// Prevent multiple instances of Prisma Client in development
declare const global: CustomNodeJsGlobal;

const prisma = global.prisma || new PrismaClient();

if (config.ENVIRONMENT === 'development') global.prisma = prisma;

export { prisma };
