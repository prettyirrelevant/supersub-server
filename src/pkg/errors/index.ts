import { type NextFunction, type Response, type Request } from 'express';
import { fromZodError } from 'zod-validation-error';
import { StatusCodes } from 'http-status-codes';
import { Prisma } from '@prisma/client';
import { ZodError } from 'zod';

import { errorResponse } from '~/pkg/responses';

export class ApiError extends Error {
  readonly status: number;

  constructor(
    status: number = StatusCodes.INTERNAL_SERVER_ERROR,
    options?: {
      message?: string;
    },
  ) {
    super(options?.message);
    this.status = status;
  }
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const handleError = (err: Error, req: Request, res: Response, _next: NextFunction) => {
  if (err instanceof ApiError) {
    if (err.status >= 500) {
      req.log.error(err, 'api error');
    }

    return errorResponse(res, err.message, err.status);
  }

  if (
    err instanceof Prisma.PrismaClientKnownRequestError ||
    err instanceof Prisma.PrismaClientUnknownRequestError ||
    err instanceof Prisma.PrismaClientRustPanicError ||
    err instanceof Prisma.PrismaClientInitializationError
  ) {
    req.log.error(err, 'database error');

    return errorResponse(
      res,
      'An unexpected database error occurred. Our team is working to resolve it.',
      StatusCodes.INTERNAL_SERVER_ERROR,
    );
  }

  if (err instanceof ZodError) {
    req.log.error(err, 'validation error');

    const validationErr = fromZodError(err);
    return errorResponse(res, validationErr.message, StatusCodes.UNPROCESSABLE_ENTITY);
  }

  req.log.error(err, 'unhandled exception');
  return errorResponse(
    res,
    'An unexpected error occurred. We are working to resolve the issue.',
    StatusCodes.INTERNAL_SERVER_ERROR,
  );
};
