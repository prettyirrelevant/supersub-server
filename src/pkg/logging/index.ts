import winston from 'winston';

import { config } from '~/pkg/env';

winston.addColors({
  warn: 'yellow',
  info: 'green',
  debug: 'cyan',
  error: 'red',
});

export const logger = winston.createLogger({
  format:
    config.ENVIRONMENT === 'development'
      ? winston.format.combine(
          winston.format.errors({ stack: true }),
          winston.format.prettyPrint(),
          winston.format.cli(),
        )
      : winston.format.combine(
          winston.format.errors({ stack: true }),
          winston.format.timestamp(),
          winston.format.metadata(),
          winston.format.json(),
        ),
  transports: [new winston.transports.Console()],
  level: config.LOG_LEVEL,
});
