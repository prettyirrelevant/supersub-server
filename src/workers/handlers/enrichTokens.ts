import { type Chain } from 'viem/chains';
import { erc20Abi } from 'viem';

import { getEvmHttpClient } from '~/pkg/evm';
import { prisma } from '~/pkg/db';

export const enrichERC20Tokens = async (chain: Chain) => {
  const tokens = await prisma.token.findMany({ where: { decimals: 0 } });

  const client = getEvmHttpClient(chain);
  const calls = tokens.map((token) => {
    const contract = { address: token.address as `0x${string}`, abi: erc20Abi } as const;

    return { ...contract, functionName: 'decimals' };
  });

  const results = await client.multicall({ contracts: calls });
  results.forEach(async (result, index) => {
    if (result.status === 'success') {
      await prisma.token.update({
        where: { address: tokens[index].address },
        data: { decimals: Number(result.result) },
      });
    }
  });
};
