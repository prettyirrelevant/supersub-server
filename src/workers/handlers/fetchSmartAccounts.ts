import { type User } from '@privy-io/server-auth';
import { type Chain } from 'viem/chains';
import { Network } from 'alchemy-sdk';

import { getMultiOwnerModularAccountAddresses, addAddressesToWebhook } from '~/pkg/evm';
import { getUniqueElements } from '~/utils';
import { logger } from '~/pkg/logging';
import { privy } from '~/pkg/privy';
import { prisma } from '~/pkg/db';

export const fetchSmartAccounts = async (chain: Chain, alchemyNetwork: Network) => {
  try {
    logger.info('Fetching Privy users');
    const privyUsers = await privy.getUsers();

    const usersByWalletAddress = privyUsers.reduce(
      (mapping, user) => {
        if (user.wallet) {
          const { wallet, ...rest } = user;
          mapping[wallet.address as `0x${string}`] = rest;
        }
        return mapping;
      },
      {} as Record<`0x${string}`, Omit<User, 'wallet'>>,
    );

    logger.info('Fetching accounts from the database');
    const accountsInDb = await prisma.account.findMany({ select: { eoaAddress: true } });
    const addressesInDb = accountsInDb.map((account) => account.eoaAddress as `0x${string}`);

    logger.info('Identifying new wallet addresses');
    const newWalletAddresses = getUniqueElements<`0x${string}`>(
      Object.keys(usersByWalletAddress) as `0x${string}`[],
      addressesInDb,
    );

    if (newWalletAddresses.length > 0) {
      logger.info('Fetching modular account addresses for new wallets', { newWalletAddresses });
      const modularAccountAddressesForNewWallets = await getMultiOwnerModularAccountAddresses(
        chain,
        newWalletAddresses,
      );

      logger.info('Creating new accounts in the database');
      await addAddressesToWebhook(Object.values(modularAccountAddressesForNewWallets), alchemyNetwork);
      await prisma.account.createMany({
        data: Object.entries(modularAccountAddressesForNewWallets).map(([walletAddress, modularAccountAddress]) => ({
          emailAddress: usersByWalletAddress[walletAddress as `0x${string}`].email?.address as string,
          metadata: { privyDid: usersByWalletAddress[walletAddress as `0x${string}`].id },
          smartAccountAddress: modularAccountAddress,
          eoaAddress: walletAddress,
        })),
        skipDuplicates: true,
      });
    }
  } catch (error) {
    logger.error(error, { description: 'Error fetching smart accounts' });
  }
};
