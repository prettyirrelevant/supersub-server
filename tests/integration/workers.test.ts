import { beforeEach, beforeAll, afterEach, describe, afterAll, expect, it, vi } from 'vitest';
import { PrivyClient } from '@privy-io/server-auth';
import { polygonAmoy } from 'viem/chains';

import { fetchSmartAccounts } from '~/workers/handlers/fetchSmartAccounts';
import { enrichERC20Tokens } from '~/workers/handlers/enrichTokens';
import { prisma } from '~/pkg/db';

describe('enrichERC20Tokens', () => {
  beforeAll(async () => {
    await prisma.token.createMany({
      data: [
        { address: '0x41E94Eb019C0762f9Bfcf9Fb1E58725BfB0e7582' }, // USDC
        { address: '0x360ad4f9a9A8EFe9A8DCB5f461c4Cc1047E1Dcf9' }, // WMATIC
        { address: '0x0Fd9e8d3aF1aaee056EB9e802c3A762a667b1904' }, // LINK
        { address: '0xcab0EF91Bee323d1A617c0a027eE753aFd6997E4' }, // CCIP BnM
      ],
    });
  });

  afterAll(async () => {
    await prisma.token.deleteMany();
    await prisma.$disconnect();
  });

  it('should update the decimals of tokens with valid results', async () => {
    const tokensBefore = await prisma.token.findMany();
    expect(tokensBefore).toHaveLength(4);

    await enrichERC20Tokens(polygonAmoy);

    const tokensAfter = await prisma.token.findMany({ select: { decimals: true, address: true } });
    expect(tokensAfter).toHaveLength(4);
    expect(tokensAfter).toContainEqual({ address: '0x41E94Eb019C0762f9Bfcf9Fb1E58725BfB0e7582', decimals: 6 });
    expect(tokensAfter).toContainEqual({ address: '0x360ad4f9a9A8EFe9A8DCB5f461c4Cc1047E1Dcf9', decimals: 18 });
    expect(tokensAfter).toContainEqual({ address: '0x0Fd9e8d3aF1aaee056EB9e802c3A762a667b1904', decimals: 18 });
    expect(tokensAfter).toContainEqual({ address: '0xcab0EF91Bee323d1A617c0a027eE753aFd6997E4', decimals: 18 });
  });
});

vi.mock('@privy-io/server-auth', async (importOriginal) => {
  const mod = await importOriginal<typeof import('@privy-io/server-auth')>();

  const PrivyClient = vi.fn();
  PrivyClient.prototype.getUsers = vi.fn();

  return { ...mod, PrivyClient };
});

describe('fetchSmartAccounts', () => {
  let privy: unknown;

  beforeEach(() => {
    privy = new PrivyClient('app-id', 'app-secret');
  });

  afterEach(async () => {
    vi.clearAllMocks();
    await prisma.account.deleteMany();
  });

  it('should create new accounts for new wallets', async () => {
    privy.getUsers.mockResolvedValueOnce(
      Promise.resolve([
        {
          wallet: { address: '0x19e4057A38a730be37c4DA690b103267AAE1d75d' },
          email: { address: 'user1@example.com' },
          id: '1',
        },
        {
          wallet: { address: '0x83fCFe8Ba2FEce9578F0BbaFeD4Ebf5E915045B9' },
          email: { address: 'user2@example.com' },
          id: '2',
        },
        {
          wallet: { address: '0xDaDC3e4Fa2CF41BC4ea0aD0e627935A5c2DB433d' },
          email: { address: 'user3@example.com' },
          id: '3',
        },
      ]),
    );

    await fetchSmartAccounts(polygonAmoy);

    const createdAccounts = await prisma.account.findMany({
      select: {
        smartAccountAddress: true,
        emailAddress: true,
        eoaAddress: true,
        metadata: true,
      },
    });
    expect(createdAccounts).toHaveLength(3);
    expect(createdAccounts[0]).toMatchObject({
      smartAccountAddress: '0x564BB9C687BE0Ff8474aD67b9D2c6a1b402A86C7',
      eoaAddress: '0x19e4057A38a730be37c4DA690b103267AAE1d75d',
      emailAddress: 'user1@example.com',
      metadata: { privyDid: '1' },
    });
  });

  it('should not create duplicate accounts for existing wallets', async () => {
    await prisma.account.create({
      data: {
        smartAccountAddress: '0x1234567890abcdef0123456789abcdef',
        emailAddress: 'user1@example.com',
        eoaAddress: '0x1234567890abcdef',
        metadata: { privyDid: '1' },
      },
    });

    privy.getUsers.mockResolvedValueOnce(
      Promise.resolve([
        {
          wallet: { address: '0x19e4057A38a730be37c4DA690b103267AAE1d75d' },
          email: { address: 'user1@example.com' },
          id: '1',
        },
        {
          wallet: { address: '0x83fCFe8Ba2FEce9578F0BbaFeD4Ebf5E915045B9' },
          email: { address: 'user2@example.com' },
          id: '2',
        },
        {
          wallet: { address: '0xDaDC3e4Fa2CF41BC4ea0aD0e627935A5c2DB433d' },
          email: { address: 'user3@example.com' },
          id: '3',
        },
      ]),
    );

    await fetchSmartAccounts(polygonAmoy);

    const createdAccounts = await prisma.account.findMany({
      select: {
        smartAccountAddress: true,
        emailAddress: true,
        eoaAddress: true,
        metadata: true,
      },
    });
    expect(createdAccounts).toHaveLength(3);
    expect(createdAccounts[1]).toMatchObject({
      smartAccountAddress: '0xeA20DB45a5D8FdF65249Eb5d22b0730aB7d2Ba9A',
      eoaAddress: '0x83fCFe8Ba2FEce9578F0BbaFeD4Ebf5E915045B9',
      emailAddress: 'user2@example.com',
      metadata: { privyDid: '2' },
    });
  });
});
