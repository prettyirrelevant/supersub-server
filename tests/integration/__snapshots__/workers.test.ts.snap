// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`indexSubscriptionPluginEvents > should index all events properly 1`] = `
[
  {
    "creatorAddress": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "description": "Share your videos with friends, family, and the world",
    "destinationChain": 80002,
    "isActive": true,
    "logoUrl": "https://t3.ftcdn.net/jpg/05/07/46/84/240_F_507468479_HfrpT7CIoYTBZSGRQi7RcWgo98wo3vb7.jpg",
    "name": "YT Nigeria",
    "onchainReference": 1,
    "receivingAddress": "0xF65330dC75e32B20Be62f503a337cD1a072f898f",
    "tokenAddress": "0x0Fd9e8d3aF1aaee056EB9e802c3A762a667b1904",
    "type": "SUBSCRIPTION",
  },
  {
    "creatorAddress": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "description": "Monthly debt repayment for eniola",
    "destinationChain": 80002,
    "isActive": true,
    "logoUrl": "https://amoy.polygonscan.com/assets/poly/images/svg/logos/logo-dim.svg?v=24.5.4.0",
    "name": "Debt repayment",
    "onchainReference": 2,
    "receivingAddress": "0xF65330dC75e32B20Be62f503a337cD1a072f898f",
    "tokenAddress": "0x0Fd9e8d3aF1aaee056EB9e802c3A762a667b1904",
    "type": "RECURRING",
  },
  {
    "creatorAddress": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "description": "A new music service with official albums, singles, videos, remixes, live performances and more for Android, iOS and desktop. It's all here.",
    "destinationChain": 80002,
    "isActive": true,
    "logoUrl": "https://music.youtube.com/img/on_platform_logo_dark.svg",
    "name": "YT Music Nigeria",
    "onchainReference": 3,
    "receivingAddress": "0xF65330dC75e32B20Be62f503a337cD1a072f898f",
    "tokenAddress": "0x0Fd9e8d3aF1aaee056EB9e802c3A762a667b1904",
    "type": "SUBSCRIPTION",
  },
]
`;

exports[`indexSubscriptionPluginEvents > should index all events properly 2`] = `
[
  {
    "chargeInterval": 3600,
    "isActive": true,
    "onchainReference": 1,
    "price": "100000000000000000",
    "productOnchainReference": 1,
  },
  {
    "chargeInterval": 9000,
    "isActive": true,
    "onchainReference": 2,
    "price": "200000000000000000",
    "productOnchainReference": 1,
  },
  {
    "chargeInterval": 2592000,
    "isActive": true,
    "onchainReference": 3,
    "price": "500000000000000000",
    "productOnchainReference": 2,
  },
  {
    "chargeInterval": 3600,
    "isActive": true,
    "onchainReference": 4,
    "price": "500000000000000000",
    "productOnchainReference": 3,
  },
]
`;

exports[`indexSubscriptionPluginEvents > should index all events properly 3`] = `
[
  {
    "creatorAddress": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "isActive": true,
    "lastChargeDate": 2024-05-29T02:12:05.000Z,
    "onchainReference": 0,
    "planOnchainReference": 3,
    "productOnchainReference": 2,
    "subscriberAddress": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "subscriptionExpiry": 2024-09-06T02:10:39.000Z,
  },
  {
    "creatorAddress": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "isActive": true,
    "lastChargeDate": 2024-05-29T02:17:19.000Z,
    "onchainReference": 1,
    "planOnchainReference": 1,
    "productOnchainReference": 1,
    "subscriberAddress": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "subscriptionExpiry": 1970-01-01T00:00:00.000Z,
  },
  {
    "creatorAddress": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "isActive": false,
    "lastChargeDate": 2024-05-29T02:20:27.000Z,
    "onchainReference": 2,
    "planOnchainReference": 4,
    "productOnchainReference": 3,
    "subscriberAddress": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "subscriptionExpiry": 2024-09-05T16:49:49.000Z,
  },
]
`;

exports[`indexSubscriptionPluginEvents > should index all events properly 4`] = `
[
  {
    "amount": "500000000000000000",
    "narration": "Recurring payment made to 0xF65330dC75e32B20Be62f503a337cD1a072f898f",
    "onchainReference": "0x87b5057d4ce7103f528ab21e5ede2931aafabb1b033e104d7e82658af303a303",
    "recipient": "0xF65330dC75e32B20Be62f503a337cD1a072f898f",
    "sender": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "status": "SUCCESS",
    "subscriptionOnchainReference": 0,
    "tokenAddress": "0x0Fd9e8d3aF1aaee056EB9e802c3A762a667b1904",
    "type": "WITHDRAWAL",
  },
  {
    "amount": "500000000000000000",
    "narration": "Recurring payment received from 0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "onchainReference": "0x87b5057d4ce7103f528ab21e5ede2931aafabb1b033e104d7e82658af303a303",
    "recipient": "0xF65330dC75e32B20Be62f503a337cD1a072f898f",
    "sender": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "status": "SUCCESS",
    "subscriptionOnchainReference": 0,
    "tokenAddress": "0x0Fd9e8d3aF1aaee056EB9e802c3A762a667b1904",
    "type": "DEPOSIT",
  },
  {
    "amount": "100000000000000000",
    "narration": "Subscription fee charged",
    "onchainReference": "0x004ea934e59d689187e41bb1852f551a9497699c2e08ca36ddad2e4825e19028",
    "recipient": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "sender": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "status": "SUCCESS",
    "subscriptionOnchainReference": 1,
    "tokenAddress": "0x0Fd9e8d3aF1aaee056EB9e802c3A762a667b1904",
    "type": "WITHDRAWAL",
  },
  {
    "amount": "100000000000000000",
    "narration": "Subscription fee received",
    "onchainReference": "0x004ea934e59d689187e41bb1852f551a9497699c2e08ca36ddad2e4825e19028",
    "recipient": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "sender": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "status": "SUCCESS",
    "subscriptionOnchainReference": 1,
    "tokenAddress": "0x0Fd9e8d3aF1aaee056EB9e802c3A762a667b1904",
    "type": "DEPOSIT",
  },
  {
    "amount": "500000000000000000",
    "narration": "Subscription fee charged",
    "onchainReference": "0x67ee590c8e4dea9e45fe869573ae36968f54a4644cc3a724cf77c88c2ef266fb",
    "recipient": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "sender": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "status": "SUCCESS",
    "subscriptionOnchainReference": 2,
    "tokenAddress": "0x0Fd9e8d3aF1aaee056EB9e802c3A762a667b1904",
    "type": "WITHDRAWAL",
  },
  {
    "amount": "500000000000000000",
    "narration": "Subscription fee received",
    "onchainReference": "0x67ee590c8e4dea9e45fe869573ae36968f54a4644cc3a724cf77c88c2ef266fb",
    "recipient": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "sender": "0x13EF7f422829C7139CEb4695762Ba87e86437e5C",
    "status": "SUCCESS",
    "subscriptionOnchainReference": 2,
    "tokenAddress": "0x0Fd9e8d3aF1aaee056EB9e802c3A762a667b1904",
    "type": "DEPOSIT",
  },
]
`;
